src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala

```diff
@@ -1,14 +1,26 @@
 package org.broadinstitute.dsde.workbench.leonardo.service
 
+import akka.http.scaladsl.model.StatusCodes
 import com.typesafe.scalalogging.LazyLogging
 import org.broadinstitute.dsde.workbench.leonardo.dao.DataprocDAO
-import org.broadinstitute.dsde.workbench.leonardo.db.DbReference
-import org.broadinstitute.dsde.workbench.leonardo.model.{Cluster, ClusterRequest, ClusterResponse}
+import org.broadinstitute.dsde.workbench.leonardo.db.{DataAccess, DbReference}
+import org.broadinstitute.dsde.workbench.leonardo.model.ModelTypes.GoogleProject
+import org.broadinstitute.dsde.workbench.leonardo.model.{Cluster, ClusterRequest, ClusterResponse, LeoException}
+import slick.dbio.DBIO
 
 import scala.concurrent.{ExecutionContext, Future}
 
+case class ClusterNotFoundException(googleProject: GoogleProject, clusterName: String) extends LeoException(s"Cluster $googleProject/$clusterName not found")
+
 class LeonardoService(gdDAO: DataprocDAO, dbRef: DbReference)(implicit val executionContext: ExecutionContext) extends LazyLogging {
 
+  def withCluster[T](googleProject: GoogleProject, clusterName: String, dataAccess: DataAccess)(op: Cluster => DBIO[T]): DBIO[T] = {
```

Comment:
this pattern is everywhere in rawls. i'm not too fond of it. any better suggestions?

Reply:
I think it's typically called the loan pattern:

https://stackoverflow.com/questions/20762240/loaner-pattern-in-scala

It's not bad by itself but in Rawls I think we tend to abuse it. :)

I question whether it's appropriate for this use case. Typically you see it when you need to "loan" access to a resource, and do some cleanup at the end (e.g. a `FileInputStream` from the SO example). Here it seems like the same thing could be done more simply:

```
  def getClusterDetails(googleProject: GoogleProject, clusterName: String): Future[Cluster] = {
    dbRef.inTransaction { dataAccess =>
      dataAccess.clusterQuery.getByName(googleProject, clusterName).map(_.getOrElse {
        throw ClusterNotFoundException(googleProject, clusterName).toErrorReport(StatusCodes.NotFound)
      })
    }
  }
```

Reply:
Another thing we could do: complete the route with a `Future[Option[Cluster]]`, and provide an unmarshaller for `Option[Cluster]` to automatically return either the 200 or 404 status :)

Reply:
So in this particular case I agree that a straight switcheroo with a call to `clusterQuery.getByName` works, but there's a future not too far from here where we'd also want to check that the user has permissions to see the cluster. That wouldn't want to roll into the db query and it might want to throw either `ClusterNotFoundException` or `ClusterPermissionException` depending on user access level*.

The advantage of the loan pattern in Rawls is that we can jam in extra checks in the one place and everyone gets them "for free".

Is the right approach here to turn `withCluster` into `getCluster(proj, name) : Try[Cluster]`, map over it, and define an unmarshaller for `Future[Try[T]]` that unpacks the exception if it exists? (riffing off your `Option` idea here)

*it wouldn't for this endpoint, because no-permission should return 404 for security reasons, but let's pretend it would

Reply:
oh, actually, s/Try/DBIO.

Reply:
```
def getCluster(proj, name, dataAccess): DBIO[Cluster] = {
    dataAccess.clusterQuery.getByName(googleProject, clusterName) map( _.match {
        case Some(cluster) => cluster
        case None => throw new ClusterNotFoundException(blah)
    })
}

...

def getClusterDetails(googleProject: GoogleProject, clusterName: String): Future[Cluster] = {
  dbRef.inTransaction { dataAccess =>
    getCluster(googleProject, clusterName, dataAccess)
  }
}
```

Reply:
^ yeah I would vote for something like that. I would even break the getting of cluster permissions outside of the DB transaction, and flatMap everything together in `getClusterDetails`. In general we should prefer flatMap over the loan pattern if possible. Because this:
```
for {
  foo <- getFoo
  bar <- getBar(foo)
  baz <- getBaz(bar)
} yield baz 
```
is infinitely better to read than:
```
withFoo { foo =>
  withBar(foo) { bar =>
    withBaz(bar) { baz => 
      baz
    }
  }
}
```

Also if we want to get into cats-land there are some things that make it easier to deal with nested contexts (like `Future[Try[A]]`. Code example:
```
  def getCluster(googleProject: GoogleProject, clusterName: String): EitherT[Future, LeoException, Cluster] = {
    val future: Future[Either[LeoException, Cluster]] = dbRef.inTransaction { dataAccess =>
      dataAccess.clusterQuery.getByName(googleProject, clusterName)
    } map {
      // Convert Option[Cluster] to Either[LeoException, Cluster]
      case Some(cluster) => Right(cluster)
      case None => Left(ClusterNotFoundException(googleProject, clusterName))
    }

    // Wrap in an EitherT
    EitherT(future)
  }

  def getClusterPermissions(cluster: Cluster): EitherT[Future, LeoException, Cluster] = {
    ??? // TODO call sam
  }

  def getClusterDetails(googleProject: GoogleProject, clusterName: String): Future[Either[LeoException, Cluster]] = {
    val transformer: EitherT[Future, LeoException, Cluster] = for {
      // flat-mapping a monad transformer unwraps _all_ the monads, giving you the inner-most thing
      cluster <- getCluster(googleProject, clusterName)
      _ <- getClusterPermissions(cluster)
    } yield cluster

    transformer.value
    // returns a Future[Either[LeoException, A]]. We could easily define an akka unmarshaller for Future[Either[LeoException, A]]
  }
```

The benefit of this is that `LeoException` is preserved in the type, and the values are easily flatmapped, despite the nested `Future[Either[..]]` context.

Reply:
I agree on flatMap > loan. On the specific case of breaking `getCluster` out of the transaction, that'd mean doing multiple transactions if you want to do e.g. `getCluster` then some other database-y thing, which is a bit sad.

I will read the monad transformer thing on a day when my head isn't broken, which today isn't.